File: Empty/library/src/main/java/com/gmail/l2t45s7e9/library/domain/ContactDetailsViewModel.java
package com.gmail.l2t45s7e9.library.domain;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.gmail.l2t45s7e9.java.entity.Contact;
import com.gmail.l2t45s7e9.java.interactor.ContactDetailsInteractor;
import com.gmail.l2t45s7e9.java.interactor.NotificationInteractor;
import com.gmail.l2t45s7e9.library.interfaces.SchedulersProvider;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import javax.inject.Inject;

public class ContactDetailsViewModel extends ViewModel {

    private ContactDetailsInteractor contactDetailsInteractor;
    private SchedulersProvider schedulersProvider;
    private NotificationInteractor notificationInteractor;
    private MutableLiveData<Contact> contactDetailsMutableLiveData = new MutableLiveData<>();
    private CompositeDisposable disposable = new CompositeDisposable();

    @Inject
    public ContactDetailsViewModel(
            ContactDetailsInteractor contactDetailsInteractor,
            NotificationInteractor notificationInteractor,
            SchedulersProvider schedulersProvider
    ) {
        this.contactDetailsInteractor = contactDetailsInteractor;
        this.notificationInteractor = notificationInteractor;
        this.schedulersProvider = schedulersProvider;
    }

    public LiveData<Contact> loadContactDetails(String id, int color) {
        disposable.add(
                contactDetailsInteractor.getContactDetailsRepo(id, color)
                        .subscribeOn(schedulersProvider.io())
                        .observeOn(schedulersProvider.ui())
                        .subscribe(contactDetailsInfo -> contactDetailsMutableLiveData.setValue(contactDetailsInfo))
        );
        return contactDetailsMutableLiveData;
    }

    public void setNotification() {
        notificationInteractor.setNotification(contactDetailsMutableLiveData.getValue());
    }

    public void cancelNotification() {
        notificationInteractor.cancelNotification(contactDetailsMutableLiveData.getValue());
    }

    public boolean getStatus() {
        return notificationInteractor.status(contactDetailsMutableLiveData.getValue());
    }

    @Override
    protected void onCleared() {
        disposable.dispose();
        super.onCleared();
    }

}

File: Empty/library/src/main/java/com/gmail/l2t45s7e9/library/domain/ContactDetailsViewModel.kt
package com.gmail.l2t45s7e9.library.domain

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import javax.inject.Inject
import com.gmail.l2t45s7e9.java.interactor.ContactDetailsInteractor
import com.gmail.l2t45s7e9.java.interactor.NotificationInteractor
import com.gmail.l2t45s7e9.java.entity.Contact
import com.gmail.l2t45s7e9.library.fragmentsState.ContactDetailsMapper
import com.gmail.l2t45s7e9.library.interfaces.DispatchersProvider
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class ContactDetailsViewModel @Inject constructor(
    private val contactDetailsInteractor: ContactDetailsInteractor,
    private val notificationInteractor: NotificationInteractor,
    private val contactDetailsMapper: ContactDetailsMapper,
    dispatchersProvider: DispatchersProvider
) : ViewModel() {

    private val contactDetailsMutableLiveData = MutableStateFlow<Contact?>(null)
    val contactDetailsLiveData = contactDetailsMutableLiveData.asStateFlow().map {
        contactDetailsMapper.create(it, status)
    }.flowOn(dispatchersProvider.default())

    fun loadContactDetails(id: String?, color: Int) = viewModelScope.launch {
        contactDetailsInteractor.getContactDetailsRepo(id, color)
            .map(contactDetailsMutableLiveData::emit)
            .collect()
    }

    fun setNotification() {
        notificationInteractor.setNotification(contactDetailsMutableLiveData.value)
    }

    fun cancelNotification() {
        notificationInteractor.cancelNotification(contactDetailsMutableLiveData.value)
    }

    private val status: Boolean
        get() = notificationInteractor.status(contactDetailsMutableLiveData.value)

}